lib/_tls_wrap.js

```diff
@@ -37,211 +37,243 @@ const tls_wrap = process.binding('tls_wrap');
+      return requestOCSPDone(new errors.Error('ERR_SOCKET_CLOSED'), self);
 
     if (response)
       self._handle.setOCSPResponse(response);
-    cb(null);
-  }
-}
-
-
-function onclienthello(hello) {
-  var self = this;
-
-  loadSession(self, hello, function(err) {
-    if (err)
-      return self.destroy(err);
+    requestOCSPDone(null, self);
+  };
 
-    self._handle.endParser();
-  });
+  self.server.emit('OCSPRequest',
+                   ctx.getCertificate(),
+                   ctx.getIssuer(),
+                   onOCSP);
 }
 
+function requestOCSPDone(err, self) {
+  if (err !== null)
+    return self.destroy(err);
 
-function oncertcb(info) {
-  var self = this;
-  var servername = info.servername;
+  if (self._handle === null)
+    return self.destroy(new errors.Error('ERR_SOCKET_CLOSED'));
 
-  loadSNI(self, servername, function(err, ctx) {
-    if (err)
-      return self.destroy(err);
-    requestOCSP(self, info, ctx, function(err) {
-      if (err)
-        return self.destroy(err);
-
-      if (!self._handle)
-        return self.destroy(new errors.Error('ERR_SOCKET_CLOSED'));
-
-      try {
-        self._handle.certCbDone();
-      } catch (e) {
-        self.destroy(e);
-      }
-    });
-  });
+  try {
+    self._handle.certCbDone();
+  } catch (e) {
+    self.destroy(e);
+  }
 }
 
 
 function onnewsession(key, session) {
-  if (!this.server)
+  const owner = this.owner;
+
+  if (!owner.server)
     return;
 
-  var self = this;
   var once = false;
-
-  this._newSessionPending = true;
-  if (!this.server.emit('newSession', key, session, done))
-    done();
-
-  function done() {
+  const done = () => {
     if (once)
       return;
     once = true;
 
-    if (!self._handle)
-      return self.destroy(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return owner.destroy(new errors.Error('ERR_SOCKET_CLOSED'));
 
-    self._handle.newSessionDone();
+    this.newSessionDone();
 
-    self._newSessionPending = false;
-    if (self._securePending)
-      self._finishInit();
-    self._securePending = false;
-  }
+    owner._newSessionPending = false;
+    if (owner._securePending)
+      owner._finishInit();
+    owner._securePending = false;
+  };
+
+  owner._newSessionPending = true;
+  if (!owner.server.emit('newSession', key, session, done))
+    done();
 }
 
 
 function onocspresponse(resp) {
-  this.emit('OCSPResponse', resp);
+  this.owner.emit('OCSPResponse', resp);
+}
+
+function onerror(err) {
+  const owner = this.owner;
+
+  if (owner._writableState.errorEmitted)
+    return;
+
+  // Destroy socket if error happened before handshake's finish
+  if (!owner._secureEstablished) {
+    // When handshake fails control is not yet released,
+    // so self._tlsError will return null instead of actual error
+    owner.destroy(err);
+  } else if (owner._tlsOptions.isServer &&
+             owner._rejectUnauthorized &&
+             /peer did not return a certificate/.test(err.message)) {
```

Comment:
Do we know if the error message is static / never changed before? In this case it would be faster to compare the whole message as string.

Reply:
Not sure and we don't have a test that covers this. I don't think we should change in this PR. I tried to avoid anything that could actually break things.